Move toMillis(Duration) to Utils so that it can be reused.

songy23 opened the pull request:
There are multiple classes that need the `toMillis(Duration)` method. I moved it to `Uitls` so that it can be reused.

songy23 commented:
Yes that could be possible, and if `Utils` ends up being too big in the future, we can split it into multiple classes, e.g StringUtils, PreconditionUtils, TimeUtils, etc.

songy23 commented:
The second approach sounds better. I can hold off this PR until we have the internal utility subproject.

sebright commented:
Here is the branch: https://github.com/sebright/opencensus-java/tree/shadow-plugin

I've never used the shadow plugin before, so I'm not sure if this is correct.  It adds a subproject with one utility method that is shared by api, impl_core, and impl.

sebright commented:
I accidentally imported the wrong Utils class in my branch, so I fixed it in a new commit.

songy23 commented:
Just found that `toMillis(Duration)` cannot be part of the internal-utils because it introduced circular dependency.

songy23 commented:
@sebright Yes there will be circular dependencies between `internal-utils` and `api` if I add `toMillis()` to internal-utils. As I see `api` already depends on `internal-utils`: https://github.com/sebright/opencensus-java/blob/shadow-plugin/api/build.gradle#L6.

sebright commented:
I was just experimenting with the shadow plugin in that branch.  The checkNotNull method in isn't needed after #1111, so the api library doesn't need to depend on it.

songy23 commented:
Got it, I was thinking about making a PR to your branch initially.

I copied internal utils and moved `toMillis()` to it instead.

bogdandrutu commented on exporters/stats/signalfx/build.gradle:
```diff
@@ -19,3 +20,10 @@ dependencies {
     signature "org.codehaus.mojo.signature:java17:+@signature"
 }
+
+shadowJar {
```
Why do we need this?

songy23 replied:
Because we don't want implementation and exporters to access opencensus-api's internal package. Instead we want to use a shaded dependency by the other artifacts, as suggested by @sebright in https://github.com/census-instrumentation/opencensus-java/pull/1114#pullrequestreview-109828652.

bogdandrutu replied:
If that is the case then it should be in common directory and implementations can use it.

songy23 replied:
Sounds reasonable, I think we can make [common.TimeUtils](https://github.com/census-instrumentation/opencensus-java/blob/master/api/src/main/java/io/opencensus/common/TimeUtils.java) public and have this method be part of it.

bogdandrutu commented:
@sebright ping on this.

sebright commented:
Sorry for the delay.  I'm fine with making toMillis public, if we add a warning about overflow.  I'd rather add it as a non-static method on Duration, though.

songy23 commented:
Sounds good, I added some java doc about the potential overflow and precision loss. I also moved `toMillis()` to `Duration.java` next to `fromMillis()`.

sebright commented on api/src/main/java/io/opencensus/common/Duration.java:
```diff
@@ -78,6 +79,20 @@ public static Duration fromMillis(long millis) {
+   * Converts a {@link Duration} to milliseconds.
+   *
+   * <p>Note that there could be overflow or loss of precision by making this conversion. See {@link
+   * TimeUnit#convert(long, TimeUnit)} for details.
```
I think it would be better to describe the overflow behavior without mentioning TimeUnit, since TimeUnit is an implementation detail, and its documentation is more general.  Also, it looks like TimeUnit.toMillis returns Long.MIN_VALUE or Long.MAX_VALUE instead of overflowing.  I don't think there is any loss of precision, besides the unavoidable conversion from nanoseconds to milliseconds.

songy23 replied:
Done - I copied the description from TimeUnit.toMillis here. I'm not sure if we should mention that converting nanoseconds to milliseconds will cause precision loss though (seems users should have known that).

sebright replied:
> I'm not sure if we should mention that converting nanoseconds to milliseconds will cause precision loss though (seems users should have known that).

I agree that we don't need to mention the precision loss.

sebright commented on api/src/main/java/io/opencensus/common/Duration.java:
```diff
@@ -78,6 +79,20 @@ public static Duration fromMillis(long millis) {
+   * @param duration a {@code Duration}.
+   * @return the milliseconds representation of this {@code Duration}.
+   */
+  public static long toMillis(Duration duration) {
```
I would rather make this method non-static, for consistency with getSeconds and getNanos.

songy23 replied:
Done.

songy23 commented on api/src/main/java/io/opencensus/common/Duration.java:
```diff
@@ -78,6 +79,20 @@ public static Duration fromMillis(long millis) {
+   * TimeUnit#convert(long, TimeUnit)} for details.
+   *
+   * @param duration a {@code Duration}.
+   * @return the milliseconds representation of this {@code Duration}.
```
Good catch, done.
